A15-binary-logging.md

```diff
@@ -0,0 +1,352 @@
+and contents of logs.
+
+The logged events correspond to what is visible at the application
+level, and as a result they exclude things like [hedging and
+retries](./A6-client-retries.md).
+
+### Log Format
+
+gRPC logs each RPC in the format defined by proto GrpcLogEntry. Initial/trailing
+metadata and messages are logged separately to match the on-the-wire
+representation.
+
+The proto of GrpcLogEntry:
+
+```
+syntax = "proto3";
+
+package grpc.binarylog.v1;
+
+import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
+
+
+// Log entry we store in binary logs
+message GrpcLogEntry {
+  // Enumerates the type of event
+  // Note the terminology is different from the RPC semantics
+  // definition, but the same meaning is expressed here.
+  enum EventType {
+    EVENT_TYPE_UNKNOWN = 0;
+    // Headers sent from client to server
+    EVENT_TYPE_CLIENT_HEADER = 1;
+    // Headers sent from server to client
+    EVENT_TYPE_SERVER_HEADER = 2;
+    // Message sent from client to server
+    EVENT_TYPE_CLIENT_MESSAGE = 3;
+    // message sent from server to client
+    EVENT_TYPE_SERVER_MESSAGE = 4;
+    // Signal that client is done sending
+    EVENT_TYPE_CLIENT_HALF_CLOSE = 5;
+    // Trailers sent from server to client.
+    // This marks the end of the RPC. Events may arrive after
+    // this due to races. The extra events are not considered
+    // a part of the RPC and may be safely ignored.
+    EVENT_TYPE_SERVER_TRAILER = 6;
+    // Signal that the RPC is cancelled.
+    // This marks the end of the RPC. Events may arrive after
+    // this due to races. The extra events are not considered
+    // a part of the RPC and may be safely ignored.
+    EVENT_TYPE_CANCEL = 7;
+  }
+
+  // Enumerates the entity that generates the log entry
+  enum Logger {
+    LOGGER_UNKNOWN = 0;
+    LOGGER_CLIENT = 1;
+    LOGGER_SERVER = 2;
+  }
+
+  EventType type = 1;
+  Logger logger = 2;  // One of the above Logger enum
+
+  // Uniquely identifies a call. Each call may have several log entries, they
+  // will all have the same call_id. Nothing is guaranteed about their values
+  // other than they are unique across different RPCs in the same gRPC process.
+  int64 call_id = 3;
+
+  // The logger uses one of the following fields to record the payload,
+  // according to the type of the log entry.
+  oneof payload {
+    // Used by EVENT_TYPE_CLIENT_HEADER, EVENT_TYPE_SERVER_HEADER and
+    // EVENT_TYPE_SERVER_TRAILER. This contains only the metadata
+    // from the application.
+    Metadata metadata = 4;
+    // Used by EVENT_TYPE_CLIENT_MESSAGE, EVENT_TYPE_SERVER_MESSAGE
+    Message message = 5;
+  }
+
+  // Peer address information, will only be recorded in
+  // EVENT_TYPE_CLIENT_HEADER or EVENT_TYPE_SERVER_HEADER
+  Peer peer = 6;
+
+  // true if payload does not represent the full message or metadata.
+  bool truncated = 7;
+
+  // The method name. Logged for EVENT_TYPE_CLIENT_HEADER
+  string method_name = 8;
```

Comment:
Hmm.  Do we also want to log the host?  I am thinking that might be useful for applications doing virtual hosting.

Is this specific to the HTTP/2 transport, or is it something we'd expect to support in other transports?

Reply:
Makes sense, `authority` is called out in the gRPC semantics doc. I'll add a `string authority`.